### ISSUE TITLE
IndexOutOfBounds exception in `FreeMarkerEmailTemplateProvider` when the IDP's displayName is an empty string

### ISSUE THREAD

**Issue by vilmosnagy** [keycloak/keycloak#9217]
### Describe the bug

When the displayName of the IDPConfig is an empty string, the IDP Linking flow fails.

Exception is:
```
13:50:35,952 WARN  [org.keycloak.services] (default task-6) KC-SERVICES0013: Failed authentication: java.lang.StringIndexOutOfBoundsException: begin 0, end 1, length 0
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)
	at java.base/java.lang.String.substring(String.java:1874)
	at org.keycloak.keycloak-common@15.1.0//org.keycloak.common.util.ObjectUtil.capitalize(ObjectUtil.java:47)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.sendConfirmIdentityBrokerLink(FreeMarkerEmailTemplateProvider.java:152)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.authenticators.broker.IdpEmailVerificationAuthenticator.sendVerifyEmail(IdpEmailVerificationAuthenticator.java:148)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.authenticators.broker.IdpEmailVerificationAuthenticator.authenticateImpl(IdpEmailVerificationAuthenticator.java:88)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.authenticators.broker.AbstractIdpAuthenticator.authenticate(AbstractIdpAuthenticator.java:76)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:446)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:276)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:389)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:253)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:389)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.continueAuthenticationAfterSuccessfulAction(DefaultAuthenticationFlow.java:186)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:164)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.authentication.AuthenticationProcessor.authenticationAction(AuthenticationProcessor.java:950)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:312)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.services.resources.LoginActionsService.brokerLoginFlow(LoginActionsService.java:829)
	at org.keycloak.keycloak-services@15.1.0//org.keycloak.services.resources.LoginActionsService.firstBrokerLoginPost(LoginActionsService.java:733)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

```

### Version

15.1.0

### Expected behavior

IDP Linking should be possible.

### Actual behavior

See the exception

### How to Reproduce?

Try to link a Google account to an existing KeyCloak account

**Comment by vilmosnagy** [comment]
@stianst any plans to merge this? We'd like to upgrade to `15.1.1` cause the security fix, but the IDP Linking flow is just broken - and I cannot even fix this from the UI, cause there's no such field on the UI as `displayName` :'(

**Comment by ross211** [comment]
> @stianst any plans to merge this? We'd like to upgrade to `15.1.1` cause the security fix, but the IDP Linking flow is just broken - and I cannot even fix this from the UI, cause there's no such field on the UI as `displayName` :'(

Looks like Keycloak doesn't set a display name and doesn't show display name as a configurable option in the UI for the Google identity provider (it does for others like SAML or OIDC). Maybe you can set it in the Keycloak db as a work around with:

    update identity_provider set provider_display_name = 'Google' where provider_alias = 'google';

**Comment by vilmosnagy** [comment]
See the linked issue (https://github.com/mrparkers/terraform-provider-keycloak/issues/639): we create this Google Identity Provider with the keycloak terraform provider, and I've found a workaround there.

But this is a bit annoying, and probably someone else will run into this. And as I can see the fix is not a huge change, should not break anything (I hope no one depends on the `ObjectUtil::capitalize` throwing an exception), so it'd be nice to merge this :)

(on the other hand, the keycloak helm chart [does not plan](https://github.com/codecentric/helm-charts/pull/515) to upgrade to the 16.x versions in the near future, and we install keycloak from a helm chart - so it won't be a solution for our case, but it could help others.)

**Comment by vilmosnagy** [comment]
@stianst @ross211 If I could do anything to accept this PR, please tell me :slightly_smiling_face:

**Comment by vilmosnagy** [comment]
fixed in c6c65ad1
